File: Filtering in MVC 3/[C#]-Filtering in MVC 3/C#/Mvc3Filter/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

using Mvc3Filter.Models;
using Mvc3Filter.Filters;
using Mvc3Filter.Helpers;

namespace Mvc3Filter.Controllers {

    public class HomeController : Controller {
        IUserDB _repository;

#if  InMemDB
        public HomeController() : this(InMemoryDB.Instance) { }
#else
        public HomeController() : this(new EF_UserRepository()) { }
#endif


        public HomeController(IUserDB repository) {
            _repository = repository;
        }

        public ViewResult Index() {
            return View("Index", _repository.GetAllUsers);
        }

        public ActionResult Details(string id) {
            User user = _repository.GetUser(id);
            if (user == null)
                return RedirectToAction("Index");

            return View("Details", user);
        }

        public ActionResult Edit(string id) {
            User user = _repository.GetUser(id);
            if (user == null)
                return RedirectToAction("Index");

            EditUserModel model = new EditUserModel();
            ModelCopier.CopyModel(user, model);

            return View(model);
        }

        [HttpPost]
        public ActionResult Edit(EditUserModel model) {
            try {
                User user = _repository.GetUser(model.UserName);
                ModelCopier.CopyModel(model, user);
                _repository.Update(user);

                return RedirectToAction("Details", new { id = model.UserName });
            } catch (Exception) {
                ModelState.AddModelError("", "Edit Failure, see inner exception.");
            }

            return View(model);
        }

        public ActionResult About() {
            return View();
        }

        public ViewResult Create() {
            return View(new CreateUserModel());
        }

        [HttpPost]
        public ActionResult Create(CreateUserModel model) {
            try {
                if (ModelState.IsValid) {
                    User user = new User();
                    ModelCopier.CopyModel(model, user);
                    _repository.CreateNewUser(user);

                    return RedirectToAction("Details", new { id = user.UserName });
                }
            } catch (Exception) {
                ModelState.AddModelError("", "Create Failure, try another user name or" +
                    " see inner exception.");
            }

            return View("Create", model);
        }

        public ActionResult Delete(string id) {
            User user = _repository.GetUser(id);
            if (user == null)
                return RedirectToAction("Index");

            return View(user);
        }

        [HttpPost]
        public RedirectToRouteResult Delete(string id, FormCollection collection) {
            _repository.Remove(id);
            return RedirectToAction("Index");
        }



        public ActionResult TestStatus(int id) {

            switch (id) {
                case 410:
                    return new HttpStatusCodeResult(410, "Please remove this link");


                case 404:
                    return new HttpStatusCodeResult(404);

                case 40404:
                    return HttpNotFound();

                case 404404:
                    return HttpNotFound("My Not Found Message");

                default:
                    return new HttpNotFoundResult("file not found");
            }
        }
    }
}

File: Filtering in MVC 3/[VB]-Filtering in MVC 3/VB/Controllers/HomeController.vb
Imports System.Web.Mvc

Imports Mvc3Filter.Models
Imports Mvc3Filter.Filters
Imports Mvc3Filter.Helpers

Namespace Mvc3Filter.Controllers

	Public Class HomeController
		Inherits Controller
		Private _repository As IUserDB

#If InMemDB Then
		Public Sub New()
			Me.New(InMemoryDB.Instance)
		End Sub
#Else
		Public Sub New()
			Me.New(New EF_UserRepository())
		End Sub
#End If


		Public Sub New(ByVal repository As IUserDB)
			_repository = repository
		End Sub

		Public Function Index() As ViewResult
			Return View("Index", _repository.GetAllUsers)
		End Function

		Public Function Details(ByVal id As String) As ActionResult
			Dim user As User = _repository.GetUser(id)
			If user Is Nothing Then
				Return RedirectToAction("Index")
			End If

			Return View("Details", user)
		End Function

		Public Function Edit(ByVal id As String) As ActionResult
			Dim user As User = _repository.GetUser(id)
			If user Is Nothing Then
				Return RedirectToAction("Index")
			End If

			Dim model As New EditUserModel()
			ModelCopier.CopyModel(user, model)

			Return View(model)
		End Function

		<HttpPost>
		Public Function Edit(ByVal model As EditUserModel) As ActionResult
			Try
				Dim user As User = _repository.GetUser(model.UserName)
				ModelCopier.CopyModel(model, user)
				_repository.Update(user)

				Return RedirectToAction("Details", New With {Key .id = model.UserName})
			Catch e1 As Exception
				ModelState.AddModelError("", "Edit Failure, see inner exception.")
			End Try

			Return View(model)
		End Function

		Public Function About() As ActionResult
			Return View()
		End Function

		Public Function Create() As ViewResult
			Return View(New CreateUserModel())
		End Function

		<HttpPost>
		Public Function Create(ByVal model As CreateUserModel) As ActionResult
			Try
				If ModelState.IsValid Then
					Dim user As New User()
					ModelCopier.CopyModel(model, user)
					_repository.CreateNewUser(user)

					Return RedirectToAction("Details", New With {Key .id = user.UserName})
				End If
			Catch e1 As Exception
				ModelState.AddModelError("", "Create Failure, try another user name or" & " see inner exception.")
			End Try

			Return View("Create", model)
		End Function

		Public Function Delete(ByVal id As String) As ActionResult
			Dim user As User = _repository.GetUser(id)
			If user Is Nothing Then
				Return RedirectToAction("Index")
			End If

			Return View(user)
		End Function

		<HttpPost>
		Public Function Delete(ByVal id As String, ByVal collection As FormCollection) As RedirectToRouteResult
			_repository.Remove(id)
			Return RedirectToAction("Index")
		End Function



		Public Function TestStatus(ByVal id As Integer) As ActionResult

			Select Case id
				Case 410
					Return New HttpStatusCodeResult(410, "Please remove this link")


				Case 404
					Return New HttpStatusCodeResult(404)

				Case 40404
					Return HttpNotFound()

				Case 404404
					Return HttpNotFound("My Not Found Message")

				Case Else
					Return New HttpNotFoundResult("file not found")
			End Select
		End Function
	End Class
End Namespace
